### DISCUSSION TITLE
why is menu positioning so restrictive?

### DISCUSSION THREAD

**Discussion by tve** [material-components/material-web#5405] [Q&A]
Why does the menu have to be positioned in the same position:relative div as the anchor without anything else? I understand the need for a common parent but this seems more restrictive than needed. Below is a link to a repro, it shows a first menu button that is in a div that has some text before the button causing the menu positioning to fail. The second menu is in a div just with the button but the div has some padding causing the positioning to fail also.

The main use-case I have is to reuse a menu (without using the positioning=document which doesn't scroll) and just keeping templates cleaner by separating the menu content from the place it's invoked.

![image](https://github.com/material-components/material-web/assets/39480/a226c717-9c76-46f7-b172-2db2228714be)

Clicking the first menu button:

![image](https://github.com/material-components/material-web/assets/39480/e40fcf7e-4421-4092-ba77-b79adf835570)

https://play.vuejs.org/#eNqtVlFv2jAQ/iteNKmtVIiAaQ80Q7RVpW1St2mbtJe8hPggbh3bsh1gQ/z3nR0SEgZrkPpC4rvvPn93Z1/YBLdK9ZcFBOMgspArnliYxCKibEmM/c3hQxwoaZhlUoyJBvSzJcQBYgiJssFksyG5WZDtNgpx5a1q8hE4l2QlNadvolCV5pz25oxzoL1ZYa0UhFEkz0EUd34dB2SacpY+o1UqEI/owY0cIAoPg2tK50Zd8x0VkiQizaQ+oPb4fUSPYbKTMksuLYIzSChnwqXmfGQQheid+J33EeeyDF+FZXSSpTa5nnnQi727ISqhlIlFz0o1HkF+Q2ZSU9BjMlBrYiRntOrvSy0bHusZGs9smqM50jVHdG6pbl+l4HevwnJ/XtvCxv3DpUk1UxbfWa6ktmRDKMyR/V7iWoCwZEvmWubkAq/vRQ27mOZIoVnCwxXMwrLwYasN/SdzEu9E+Z9OIJ9RiYwFrD0WVSYF98+m2suNS5wmNrm8Iv6d4AmwhRbVirhBMiZx8Ljbi/yCGbFgbBxckxKzdY/ttfvNwWaSmnEVXh2/S6j5CbEZM/23eNJM3ydlMrm6vNpxeZp94PB/kcMjoShmi4Yo3LUquA7KevXyRGFVpMCp6gnjncPEQa03DjjDzFzCmbXKjMMwpa43FDhb6r4AGwqVh4iajkImKKzRiaVohIed4xtx7V52YGgFTActLieCyzTh7A+e+5epmngnrVct2rl5nIYkdROrBxywB12KdTQOJR+aOuzWqTIntmt3qWZsvHsBXXpXs777J7y1R2Zz3pHRQfFIVK/HuToerYqswWANZ7MO0R43HZbPPpj3/dwpqW4W3iZrUinmbHFwl1KcKYyD/qrcl619pxL847H67G1WF1CLSjNIn4/Yn8y6lPpNgwHt/trsE0n0Anan7uHHF1i7QVQ5c0kLvjvyJ5zfAT+mhdNYwu4KQVF2A+fVfvKjAT/IP83D2oIwVVJO6H7exQHOeTdOT6W+lzvqj+oqbv8CGS5YPg==

Update: using "positioning=document" doesn't work well either, try it on the second menu button in the repro: the menu shows up way down the page.

**Comment by asyncliz** [comment]
@e111077 may have some insights here.

I'm curious though, can you not separate the menu template from the button in Vue? For example, in Lit it could look like

```js
html`
  <div style="position: relative">
    <md-filled-button id="anchor">Anchor</md-filled-button>
    ${getMenuFor('anchor')}
  </div>
`
```

**Reply by tve** [reply]
I'm not familiar with Lit so I don't know what the exact semantics are of the snippet you posted. I can obviously put the menu in another component but that results in a proliferation of names and files as opposed to having it just a few lines down in the template. There's also a question about how many components get instantiated when (think 1000 line table with an actions menu button).

**Reply by asyncliz** [reply]
Ah I understand, so it's about sharing a single instance of a menu between multiple anchors?

**Reply by tve** [reply]
Yes

**Comment by e111077** [comment]
small note, try positioning="popover" which should work for most cases. Our API docs bot is broken so the API docs are a bit out of date on menu

## The problems

There are several reasons as to why positioning is so restrictive – there's just no one-size-fits-all solution as of Jan 2024.

1. We do not know the context of the application that it is being used in
2. We don't want to break out of the DOM tree's rendering for styling and events
3. Performance

### We don't know your app

We want menu to render above all other elements on the page. Standard practice here has been to detach the menu from the page and render it at the end of the document. Then one would typically listen for scrolling and reposition the element based on the scroll. The problem is that we do not know if it is safe to remove the element from the page and we do not know if it is safe to render the element at the end of `<body>` because we do not know your application or how your renderer works. If we remove the element from the page what is stopping your unknown renderer from recreating the element that was just removed or deleting the new position of the element. What if we are in a server context? then you have to be careful about calling DOM operations

If we could safely assume this, then we know what element to listen to for scrolls (body) since scroll events don't bubble and thus we can do repositioning on scrolling.

### Keeping the DOM tree

We don't want to move elements out of the DOM tree, because then you lose the powers of the DOM tree such as event propagation and style application. We use events heavily in menus with submenus and people often style their components with CSS custom properties which cascade down the tree. This would almost immediately cause problems with our `<md-select>` element's styles. We'd have to then find a way to also re-propagate events and pass styles with menu's new location

### Performance

Even if we were able to hoist the element to the bottom of the page and do repositioning-on-scroll, the actual reposition calculation is expensive and requires at least 2 RAFs which is incredibly slow and becomes unscalable when scrolling with multiple submenus.

With position absolute, position fixed, and position document (which is just absolute calculated from the document rather than from the anchor) we either leave the hard calculations to the browser engine or just ignore them in the case of positioning fixed.

## The solutions / future

We have a plan for this

### current solution

Offer users a lot of customizability to select their preferred rendering mode to fit their situation. e.g. sometimes `positioning=fixed` is best since it breaks out of stacking context in many cases, but it doesn't scroll with the page. `positioning=absolute` is good if you can control what is `position:relative` above it, and `positioning=document` is good if you can hoist the menu yourself or if you want to reuse the menu element and render it on demand.


### future solution

we have lately been experimenting with **`positioning=popover`** which uses the native [popover API](https://developer.mozilla.org/en-US/docs/Web/API/Popover_API) which will host to the [top-layer](https://developer.mozilla.org/en-US/docs/Glossary/Top_layer). On unsupported browsers it will default to `positioning=fixed`. The good thing about popover is that it will render above everything and scroll with most contexts. The problem is that sometimes it may not scroll with an element which is not scrolling with the rest of the document (e.g. the anchor is in a scrollable pane).

Because of performance, we still do not offer repositioning, but we hope to eventually make `positioning=popover` the only positioning option once we have it across browsers and have the [native browser CSS anchor positioning](https://developer.chrome.com/blog/tether-elements-to-each-other-with-css-anchor-positioning) shipped in at least one browser with likely a ponyfill for other browsers. I have been working with the web platform teams at google to make sure that this proposal fits our use case. It is also somewhat usable now in Chrome Canary with the Experimental Web Features flag on and it has some promising performance but still missing some things such as passing the anchor reference as a CSS Custom Property.
